File: backend/app/schemas/company.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional, Literal, Any
from uuid import UUID

class CompanyBase(BaseModel):
    name: str
    role: str
    status: Literal['wishlist', 'applied', 'interviewing', 'offered', 'rejected'] = "wishlist"
    applied_date: Optional[datetime] = None
    interview_date: Optional[datetime] = None
    job_url: Optional[str] = None
    salary_range: Optional[str] = None
    notes: Optional[str] = None
    
    # Intelligence metadata
    industry: Optional[str] = None
    tier_category: Optional[str] = None
    difficulty: Optional[str] = None
    preparation_topics: Optional[list[str]] = None
    source_metadata: Optional[dict[str, Any]] = None

    # Logo.dev integration & official brand domain fields
    official_domain: Optional[str] = None
    logo_provider: Optional[str] = "logo.dev"
    logo_status: Optional[str] = "UNVERIFIED"

class CompanyCreate(CompanyBase):
    pass

class CompanyUpdate(CompanyBase):
    name: Optional[str] = None
    role: Optional[str] = None
    status: Optional[Literal['wishlist', 'applied', 'interviewing', 'offered', 'rejected']] = None
    applied_date: Optional[datetime] = None
    interview_date: Optional[datetime] = None
    job_url: Optional[str] = None
    salary_range: Optional[str] = None
    notes: Optional[str] = None
    industry: Optional[str] = None
    tier_category: Optional[str] = None
    difficulty: Optional[str] = None
    preparation_topics: Optional[list[str]] = None
    source_metadata: Optional[dict[str, Any]] = None
    official_domain: Optional[str] = None
    logo_provider: Optional[str] = None
    logo_status: Optional[str] = None

class CompanyResponse(CompanyBase):
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

class CompanyDirectoryItem(BaseModel):
    rank: int
    name: str
    normalized_key: str
    aliases: list[str] = []
    industry: Optional[str] = None
    tier_category: Optional[str] = None
    score: int
    in_both_datasets: bool
    questions_dataset_1: int
    questions_dataset_2: int
    total_problem_references: int
    overall_difficulty: str
    difficulty_breakdown: dict[str, int]
    top_preparation_topics: list[str]
    hiring_recency_windows: list[str]
    hiring_frequency: str
    job_roles: list[str] = []
    salary_range: Optional[str] = None
    interview_rounds: Optional[str] = None
    sources: list[str] = []
    official_domain: Optional[str] = None
    logo_provider: Optional[str] = "logo.dev"
    logo_status: Optional[str] = "UNVERIFIED"

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/services/api/company.ts
import { apiClient } from '@/services/api/client';

export interface Company {
  id: string;
  name: string;
  role: string;
  status: 'wishlist' | 'applied' | 'interviewing' | 'offered' | 'rejected';
  appliedDate?: string;
  interviewDate?: string;
  jobUrl?: string;
  salaryRange?: string;
  notes?: string;
  industry?: string;
  tierCategory?: string;
  difficulty?: string;
  preparationTopics?: string[];
  sourceMetadata?: any;
  createdAt: string;
  updatedAt: string;
}

export interface CompanyDirectoryItem {
  rank: number;
  name: string;
  normalized_key: string;
  aliases: string[];
  industry?: string;
  tier_category?: string;
  score: number;
  in_both_datasets: boolean;
  questions_dataset_1: number;
  questions_dataset_2: number;
  total_problem_references: number;
  overall_difficulty: string;
  difficulty_breakdown: Record<string, number>;
  top_preparation_topics: string[];
  hiring_recency_windows: string[];
  hiring_frequency: string;
  job_roles: string[];
  salary_range?: string | null;
  interview_rounds?: string | null;
  sources: string[];
}

export type CompanyCreate = Omit<Company, 'id' | 'createdAt' | 'updatedAt'>;
export type CompanyUpdate = Partial<CompanyCreate>;

export interface CompanyQueryParams {
  search?: string;
  status?: 'wishlist' | 'applied' | 'interviewing' | 'offered' | 'rejected';
  role?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  skip?: number;
  limit?: number;
}

export const companyApi = {
  getCompanies: async (params?: CompanyQueryParams): Promise<Company[]> => {
    const { data } = await apiClient.get('/companies', { params });
    return data?.data ?? data;
  },

  getCompany: async (id: string): Promise<Company> => {
    const { data } = await apiClient.get(`/companies/${id}`);
    return data?.data ?? data;
  },

  getDirectory: async (search?: string): Promise<CompanyDirectoryItem[]> => {
    const { data } = await apiClient.get('/companies/directory', { params: { search } });
    return data?.data ?? data;
  },

  createCompany: async (company: CompanyCreate): Promise<Company> => {
    const { data } = await apiClient.post('/companies', company);
    return data?.data ?? data;
  },

  updateCompany: async (id: string, company: CompanyUpdate): Promise<Company> => {
    const { data } = await apiClient.put(`/companies/${id}`, company);
    return data?.data ?? data;
  },

  deleteCompany: async (id: string): Promise<void> => {
    await apiClient.delete(`/companies/${id}`);
  },

  bulkDeleteCompanies: async (ids: string[]): Promise<void> => {
    await apiClient.post('/companies/bulk-delete', ids);
  },
};
